---
title: >
  The `CIA` User's Guide
author:
- name: Ivan Ferrari
  affiliation: 
  - Istituto Nazionale Genetica Molecolare (INGM), Milan, Italy
  email: ferrari@ingm.org
- name: Andrea Gobbini
  affiliation: 
  - Istituto Nazionale Genetica Molecolare (INGM), Milan, Italy
  email: gobbini@ingm.org
- name: Mattia Battistella
  affiliation: 
  - Istituto Nazionale Genetica Molecolare (INGM), Milan, Italy
  email: battistella@ingm.org
- name: Eugenia Galeota
  affiliation: 
  - Istituto Nazionale Genetica Molecolare (INGM), Milan, Italy
  email: galeota@ingm.org
- name: Federico Marini
  affiliation: 
  - Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  email: marinif@uni-mainz.de
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('CIAr')`"
output: 
  BiocStyle::html_document:
    toc_float: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{The CIA User's Guide}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{CIA}
  %\VignetteKeywords{Classification, GeneExpression, SingleCell, Transcriptomics,
    Annotation, Clustering, Visualization}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: CIA_bibliography.bib
---

<style type="text/css">
.smaller {
  font-size: 10px
}
</style>

**Compiled date**: `r Sys.Date()`

**Last edited**: 2024-07-11

**License**: `r packageDescription("CIAr")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  error = FALSE,
  warning = FALSE,
  eval = TRUE,
  message = FALSE,
  fig.width = 10
)
options(width = 100)
```

<hr>

# Introduction {#introduction}

**Cluster Independent Annotation** (**CIA**) is a classification method designed to **assist researchers during the cell annotation step** of scRNA-seq experiments. 
Given gene signatures as input, this classifier **computes a signature score for each cell** and **compares the score values** to assign a label to each single cell. 

This tool offers several advantages:

-	It synthesizes the information of **an entire signature expression into a single score value**, avoiding the tedious inspection of individual marker genes from lengthy differentially expressed genes (DEGs) lists, which may not be cluster-specific individually.

- It makes possible the exploration of new signatures derived from your research for which training data for machine learning algorithms may not yet exist.

-	It provides a classification for each cell that is **completely independent of clustering**, allowing it to be used alongside a clustering method to set a proper resolution value, thus yielding coherent and easily annotated cell groups.

-	**It's very fast**: it can classify a large dataset (hundreds of thousands of cells) in just a few seconds since **we have implemented the ability to parallelize processes**.

-	Being signature-based, this tool can deliver insights on **any kind of biologically meaningful gene list**, also enabling functional annotation. 

-	By normalizing for the gene signature length, it facilitates the **comparison of gene sets of varying lengths**, ranging **from tens to thousands of genes**.

# Getting started {#gettingstarted}

To install this package, start R and enter:

```{r install, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("CIAr")
```

Once installed, the package can be loaded and attached to your current workspace as follows:

```{r loadlib, eval = TRUE}
library("CIAr")
```

If you want to install the development version (from the GitHub repository), you should instead run the following line:

```{r installdev, eval=FALSE}
BiocManager::install("ingmbioinfo/CIA_R")
```

# CIA input {#inputtocia}

Our method requires as input a **named list of gene signatures** (e.g. cell type, cell state ...) and a scRNA-seq data in **[SeuratObject](https://rdrr.io/cran/SeuratObject/) or [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) format**.

## Gene signatures {#inputgenesignatures}

In our study, we used the differentially expressed genes (DEGs) from Hao *et al.*'s **PBMC atlas** as signature markers [@Hao2021]. 
The original clusters in this dataset were annotated using an integrated analysis combining RNA and protein data, ensuring the **RNA-based gene lists accurately represent specific cell types**. 
We focused on the broadest annotation level for clearer visualization and easier cross-dataset comparison.
**We omitted the 'other T'** label from our analysis, as it includes cell types not present in the PBMC3K dataset, making validation impossible. 
The **'other'** cluster, predominantly platelets, **was relabeled as 'Platelet'** for clarity.  
N.B. - DEGs have been filtered to create concise and targeted gene lists: a log2 fold change greater than 1.5, a minimum average expression level of 0.25, a z-score above 5, and expression in at least 40% of the cells within each cluster.

We start our analysis by loading the required packages:

```{r loadpkgs}
library("CIAr")
library("SingleCellExperiment")
library("Seurat")
library("Matrix")
library("sparseMatrixStats")
library("igraph")
library("patchwork")
```

Subsequently, we load the `atlas.gmt` file into our environment and inspect it briefly.

```{r loadgmt}
## To load the gene signatures
gmt <- load_signatures(system.file("extdata", "atlas.gmt", package = "CIAr"))

# One can load_signatures can load both dictionaries or gmt files by providing both file_paths or URLs

for (i in names(gmt)) {
  cat(paste0(i, ":"), length(gmt[[i]]), "genes \n")
}
```

The `signatures_similarity()` function is used to check the similarity among the gene lists (using the Jaccard Index by default)

```{r similgmt}
signatures_similarity(gmt, metric_name = "jaccard")
```

## Data input {#inputscdata}

In order **to evaluate both the consistency** of our method **and the performances of classification**, we used the PMBC atlas DEGs to automatically annotate the [PBMC3K](https://scanpy.readthedocs.io/en/stable/generated/scanpy.datasets.pbmc3k.html) dataset from [@Satija2015]. 
This dataset was annotated by the authors relying on clustering and marker genes expression inspection and it is widely used as reference in the scientific community.
**We classified** this dataset **independently from the already present annotation**, whose **cell labels were used as ground truth** to evaluate our classification performances with different modalities.

Our method support both SeuratObject (SO) an SingleCellExperiment objects as input (SCE), and it is tested on expression data previously scaled to 10000 and Log-normalized; for this reason we strongly suggest to normalize your count matrix as follows.

For a `Seurat` object:

```{r loadso}
SO <- readRDS(system.file("extdata", "pbmc3k.rds", package = "CIAr"))
SO <- NormalizeData(SO, 
                    normalization.method = "LogNormalize", 
                    scale.factor = 10000) ## To add "data" layer
SO
```

For a `SingleCellExperiment` object, we can use the following chunk to obtain normalized expression values in the same flavour provided by the `scanpy` framework:
 
```{r loadsce}
SCE <- as.SingleCellExperiment(SO)
library_sizes <- colSums(assay(SCE, "counts"))
normalized_counts <- sweep(assay(SCE, "counts"), 2, library_sizes, FUN = "/") * 10000
assay(SCE, "logcounts") <- log1p(normalized_counts) 
SCE
```

```{r graphsce}
SCE@metadata$graphs[["RNA_snn"]] <- SO@graphs$RNA_snn

# TODO: We might need to re-think how to store this, and likely use accessors like
# metadata(SCE) at least for the first part
```

For the classification of the test dataset **we renamed and merged some clusters** in order to make easier the comparison and the visualization of results.
In particular, ‘CD14+ Monocytes' and 'FCGR3A+ Monocytes' clusters were merged into ‘Mono’. 

```{r firstumap, fig.width=14, fig.height=7}
DimPlot(SO, group.by = "seurat_annotations") | 
  DimPlot(SO, group.by = "celltype", cols = c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728",
                                              "#6fadfd", "#E4D00A", "#FF5733")
  )
```

# Signature score {#sigscore}

The `score_signature()` function is **based on the "gene signature score"** calculation method presented in [@DellaChiara2021].  
The "gene signature score" of a signature **S** in a cell **j** is defined as follows:

$${GSS}(S, j) = \left( \frac{n_j}{L} \right) \times \left( \frac{\sum_{i \in S \cap G_j} X_{ij}}{\sum_{i \in G_j} X_{ij}} \right)$$
 
Where:

- $n_j$ is the number of genes in the signature $S$ that have non-zero expression in cell $j$
- $L$ is the total number of genes in the signature $S$
- $S$ is the set of genes in the signature
- $𝐺_j$ is the set of genes expressed (i.e. with counts > 0) in cell $j$
- $X$ is the gene expression matrix where $X_{ij}$ represents the expression of gene $i$ in cell $j$

With this score it is possible to **condensate in a single value both the proportion of expressed signature genes and their overall expression**, enabling researchers to easily study whole signatures expression at single cell level. 

Here we show how to compute the scores of a gene signature ("B cells", named `B` in the `gmt` object), for both a SeuratObject and a SingleCellExperiment object:

```{r bcellsscores}
## Compute the signature scores of B cells gene signature for each cell
head(score_signature(data = SO, geneset = gmt[["B"]]))

head(score_signature(data = SCE, geneset = gmt[["B"]]))
```

## Default score {#scoredefault}

To compute the signature scores of each signature in parallel, we implemented the `score_all_signatures()` function.
The function calls `load_signatures`, allowing it to support both **named lists** and canonical **gmt files** (tab separated and without header), which can be provided as either a **filepath** or a **URL**.

For **SeuratObjects**, it is possible to select the matrix on which to compute the scores by setting the `seurat_assay` and `seurat_layer` parameters. By default, these are set to `RNA` and `data`, respectively.

```{r so-scoreall}
SO <- score_all_signatures(SO,
                           signatures_input = gmt,
                           seurat_assay = "RNA", 
                           seurat_layer = "data", 
                           n_cpus = 2)

head(SO@meta.data[, names(gmt)])
```

For **SingleCellExperiment** objects, the same can be done by setting the `sce_assay` parameter, which is set to `logcounts` by default.

```{r sce-scoreall}
SCE <- score_all_signatures(SCE, 
                            signatures_input = gmt, 
                            sce_assay = "logcounts", 
                            n_cpus = 2)

head(colData(SCE)[, names(gmt)])
```

From this point in the tutorial, **we will demonstrate CIA features using data in SeuratObject format** for clarity purposes. 
The following parameter settings work exactly the same for SingleCellExperiment objects.

## Scaled score {#scorescaled}

The scaled score is the **GSS divided by the maximum score value**, an operation that rescales the values from 0 to 1. 
This allows scaled scores of **different signatures**, even with **varying lengths**, to be **directly compared**. To compute the 'scaled' score, score_mode must be set to 'scaled'.

```{r scoreallseurat, fig.width=8, fig.height=7}
SO <- score_all_signatures(data = SO, 
                           signatures_input = gmt, 
                           score_mode = "scaled", 
                           n_cpus = 2)

head(SO@meta.data[, names(gmt)])

DimPlot(SO, 
        group.by = "celltype", 
        cols = c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
                 "#6fadfd", "#E4D00A", "#FF5733"))
```

By inspecting the score values, for all the signatures, **the highest values are found in the proper cluster**, indicating the sensitivity of the signatures and the capability of the signature score to represent the expression of the whole gene lists.

```{r plotallscores, fig.width=15, fig.height=10}
FeaturePlot(object = SO, 
            features = names(gmt), 
            cols = c("lightgrey", "red"))
```

# Signature-based classification {#classifyfull}

To better visualize those distributions, we developed the `grouped_distributions()` function. 
By providing cell/sample metadata that include the signature scores, this function plots a heatmap showing the medians of their values in cell groups and it prints a brief statistical report.  
**For each cell group**, a two-sided **Wilcoxon test** is performed to evaluate if the distribution with the highest median is different from the others. 
**For each signature**, a two-sided **Mann-Whitney U test** is performed to evaluate if the distribution in the cell group having the highest median is different from the other groups distributions.

```{r groupeddistros, fig.width=10, fig.height=8}
grouped_distributions(SO@meta.data, 
                      ref_obs = "celltype", 
                      columns_obs = names(gmt), 
                      scale_medians = "column-wise")
```

The statistical tests confirmed that the **visible differences in signature score distributions are significant**, indicating that scaled signature scores are consistent with authors annotation. 
With the evidence of the goodness of the signatures, we proceeded with the classification of PBMC3K.

## Default classification {#classifydefault}

To classify the PBMC3K dataset we use the `CIA_classify()` function, which directly computes and compares scaled scores of each signature in each single cell of the dataset.

```{r classifyso}
SO <- CIA_classify(data = SO, 
                   signatures_input = gmt,
                   similarity_threshold = 0, 
                   column_name = "CIA_default", 
                   n_cpus = 2)
```

Classification is performed by **assigning to each cell the label of the signature that has the maximum scaled score value**. 
Because it is based on matrix and vector operations, and given the possibility to parallelize the computation, this process is **very fast** and lends itself to large datasets and different iterations for an optimal definition of the cell identities.

```{r plotclass, fig.width=14, fig.height=7}
cols <- c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
          "#6fadfd", "#E4D00A", "#FF5733")
DimPlot(SO, group.by = "celltype", cols = cols) | 
  DimPlot(SO, group.by = "CIA_default", cols = cols)
```

```{r groupdistro, fig.width=10, fig.height=8}
group_composition_heatmap(SO@meta.data, 
                          classification_obs = "CIA_default", 
                          ref_obs = "celltype")
```

The confusion matrix reveals that the classification method is highly effective, with most cells being correctly labeled according to their true cell types. 
The errors that did occur are relatively few and can be summarized as follows:

- **CD4 T cells:** Predominantly classified correctly, with a small number misclassified as CD8 T cells.
- **CD8 T cells:** Mostly identified accurately, but with some confusion with NK cells noted.
- **Dendritic cells:** Generally well-classified, although there is a tendency for some cells to be mislabeled as monocytes.

## Classification with similarity threshold {#classifythreshold}

The similarity threshold is a critical parameter in the classification process, utilized within the `CIA_classify()` function. 
It sets **the minimum required difference between the highest and second-highest signature scores** for a cell to be confidently classified into a specific category. 
This threshold **prevents ambiguous classifications** and ensures that cells are distinctly assigned to the most appropriate categories based on their signature expression levels.

Cells whose signature scores do not meet this threshold are labeled as **'Unassigned'**. 
This label indicates that these cells do not exhibit a strong correlation with any of the predefined signatures, thus avoiding misclassification.

Here we test 3 thresholds: 5%, 10% and 15% - the classification values are stored each in a separate vector, added to the original object.

```{r classify-unassigned, fig.width=12, fig.height=7}
for (t in c(0.05, 0.1, 0.15)) {
  SO <- CIA_classify(
    data = SO, signatures_input = gmt, unassigned_label = "Unassigned",
    similarity_threshold = t, column_name = paste0("CIA_t", as.character(t)), n_cpus = 2
  )
}

cols <- c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
          "#6fadfd", "#E4D00A", "#FF5733")
DimPlot(SO, group.by = "celltype", cols = cols) | 
  DimPlot(SO, group.by = "CIA_default", cols = cols)
```

```{r plotthreetresholds, fig.width=18, fig.height=5}
cols <- c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
          "#6fadfd", "#E4D00A", "#FF5733", "#808080")
DimPlot(SO, group.by = "CIA_t0.05", cols = cols) | 
  DimPlot(SO, group.by = "CIA_t0.1", cols = cols) |
  DimPlot(SO, group.by = "CIA_t0.15", cols = cols)
```

Raising the similarity threshold in the classification algorithm indeed leads to a greater number of cells being labeled as **"Unassigned"**. 
This change typically affects **cells that previously might have been incorrectly classified**, particularly those within groups that exhibit closely related signature profiles. 
As the similarity threshold increases, the classifier becomes more conservative, requiring a higher degree of confidence before assigning a label. 
Consequently, **cells that do not distinctly match a particular signature according to the stricter criteria are more likely to remain unclassified**.

To further show the importance of this threshold we removed a signature from the dictionary in order to simulate the situation in which a cell type of the dataset is not represented by the signatures.

```{r bgone}
gmt$B <- NULL
```

After eliminating the B cell signature, we re-run the classification, again storing the results in a specific vector added to the original object.

```{r classifynob, fig.width=6, fig.height=5}
for (t in c(0, 0.1)) {
  SO <- CIA_classify(
    data = SO, signatures_input = gmt, unassigned_label = "Unassigned",
    similarity_threshold = t, column_name = paste0("CIA_noB_t", as.character(t)), n_cpus = 2
  )
}

cols <- c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
          "#6fadfd", "#E4D00A", "#FF5733")
DimPlot(SO, group.by = "celltype", cols = cols)
```

```{r plotnob, fig.width=12, fig.height=5}
cols <- c("#b2df8a", "#1f78b4", "#d62728", "#6fadfd", 
          "#E4D00A", "#FF5733", "#808080")
DimPlot(SO, group.by = "CIA_noB_t0", cols = cols) | 
  DimPlot(SO, group.by = "CIA_noB_t0.1", cols = cols)
```

Indeed, applying a similarity threshold has proven to be a **crucial step for enhancing the specificity and reliability of the classification results**. 
It has effectively reduced the mislabeling of B cells as Dendritic Cells (DC), demonstrating its value in refining the accuracy of cell type predictions. 
Moreover, the threshold's role in identifying 'Unassigned' cells **helps flag potential new or unexpected cell types**, such as contaminants or previously unidentified cell populations, within the sample. 
This feature is especially beneficial for exploratory analyses, where novel discoveries are a possibility.

# Classification performance evaluation {#performanceeval}

To evaluate classification performances both per ground truth cluster and overall we exploited respectively `grouped_classification_metrics()` and `compute_classification_metrics()` functions.  

In both functions **cell labels assigned by CIA and the annotation already present in test datasets are compared** in order to count true positive (TP), true negative (TN), false positive (FP) and false negative (FN) cells for each cluster. 
Only for the overall calculation the per-cluster counts are summed to obtain the total TN, TP, FN and FP.

Then, again for both functions, the following metrics are calculated:

- **Sensitivity**: $SE = TP/(TP + FN)$
- **Specificity**: $SP = TN/(TN + FP)$
- **Precision**: $PR = TP/(TP + FP)$
- **Accuracy**: $ACC = (TN + TP)/(TN + TP + FN + FP)$
- **F1-score**: $F1 = 2 \times TP/(2 \times TP + FN + FP)$

N.B.: the column of the classification of interest and the one with ground truth labels must have the same categories to be compared.  
Here, for clarity, we show only the **per-cluster classification** metrics of a single classification.

```{r gcmetrics}
grouped_classification_metrics(SO@meta.data, 
                               classification_col = "CIA_default", 
                               ref_labels = "celltype")
```

And here are report the **overall performances of each classification modality**:

```{r allmetrics}
cols <- c("CIA_default", "CIA_t0.05", "CIA_t0.1", 
          "CIA_t0.15", "CIA_noB_t0", "CIA_noB_t0.1")
compute_classification_metrics(SO@meta.data, 
                               classification_cols = cols, 
                               ref_labels = "celltype")
```

As the similarity threshold is raised, the **overall classification performance may decrease due to an increase in the number of 'Unassigned' cells**, which are counted as misclassified. 
However, by setting the `unassigned_label` parameter to the specific value, we can exclude 'Unassigned' cells from performance metrics calculations, thereby refining our analysis. 
Consequently, a new column indicating the percentage of unlabeled cells will be included in the output, providing additional insight into the classification process.

```{r gcmunassigned}
compute_classification_metrics(SO@meta.data,
                               classification_cols = cols, 
                               ref_labels = "celltype",
                               unassigned_label = "Unassigned")
```

# Additional functionality of CIA {#additionalfuncs}

In this section, we display the additional functionality provided by the CIA package.

## Direct inspection of MSigDB signatures {#addmsigdb}

With the `score_signature()` and `score_all_signatures()` functions, you can compute and inspect signatures from the [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/) by simply providing the corresponding URL.

```{r checkmsigdb}
phago_signature <- 
  load_signatures("https://www.gsea-msigdb.org/gsea/msigdb/human/download_geneset.jsp?geneSetName=GOBP_PHAGOCYTOSIS&fileType=gmt")

head(
  score_signature(
    SO,
    geneset = phago_signature$GOBP_PHAGOCYTOSIS
  )
)
```

```{r scoregmt}
SO <- score_all_signatures(SO,
  score_mode = "scaled",
  signatures_input = "https://www.gsea-msigdb.org/gsea/msigdb/human/download_geneset.jsp?geneSetName=GOBP_PHAGOCYTOSIS&fileType=gmt", 
  n_cpus = 2
)
```

```{r plotgobp, fig.width=8, fig.height=6}
FeaturePlot(object = SO, 
            features = "GOBP_PHAGOCYTOSIS", 
            cols = c("lightgrey", "red"))
```

## Majority voting {#addmajority}

The **Celltypist** classifier [@DomnguezConde2022] features an intriguing process called **'majority voting,'** which refines cell identities within subclusters following an **over-clustering step**. 
In essence, **within each subcluster**, the **label of the predominant cell type is applied** to all cells in that group. 
While this step extends beyond the standard CIA workflow, we aimed to compare classification results under optimal conditions. 
To this end, we developed the `celltypist_majority_vote()` function to emulate the 'majority voting' mechanism.

N.B.: Due to a different implementation of the Leiden algorithm, results may be slightly different from the Python version of CIA.

```{r majvote}
SO <- celltypist_majority_vote(SO, 
                               classification_obs = c("CIA_default", "CIA_t0.1"))
```

```{r plotoc, fig.width=10, fig.height=6}
DimPlot(SO, group.by = "overclustering")
```

```{r fig.width=12, fig.height=5}
cols1 <- c("#b2df8a", "#fb9a99", "#6fadfd", "#E4D00A", 
           "#1f78b4", "#d62728", "#FF5733")
cols2 <- c("#b2df8a", "#fb9a99", "#6fadfd", "#E4D00A", 
           "#1f78b4", "#808080", "#d62728", "#FF5733")
DimPlot(SO, 
        group.by = "CIA_default_majority_voting", 
        cols = cols1) | 
  DimPlot(SO, 
          group.by = "CIA_t0.1_majority_voting", 
          cols = cols2)
```

```{r classmetmv}
cols <- c("CIA_default_majority_voting", "CIA_t0.1_majority_voting")
compute_classification_metrics(SO@meta.data, 
                               classification_cols = cols, 
                               ref_labels = "celltype", 
                               unassigned_label = "Unassigned")
```

## Extraction of signatures from Differentially Expressed Genes {#adddesigs}

With Seurat, it is possible to extract signatures to be used within CIA to classify other datasets in just a few steps:

```{r designatures}
Idents(SO) <- SO@meta.data$celltype
pbmc_degs <- FindAllMarkers(SO, logfc.threshold = 0.5, min.pct = 0.3, only.pos = TRUE)
pbmc_degs <- pbmc_degs[pbmc_degs$p_val < 0.01, ]
pbmc_gmt <- list()

for (i in unique(pbmc_degs$cluster)) {
  pbmc_gmt[[i]] <- pbmc_degs[pbmc_degs$cluster == i, "gene"]

  cat(paste0(i, ":"), length(pbmc_gmt[[i]]), "genes \n")
}
```

## Alternative representation of cell group composition {#addvisualsummaries}

Alternatively to the canonical confusion matrix, it is possible to visualize the classification results using the `plot_group_composition()` function. 
This function shows the composition of each reference group given the newly performed classification, allowing the use of the same colors used for the corresponding UMAP plot.

```{r plotgc}
# Plot group composition
plot_group_composition(SO@meta.data,
                       comp_col = "CIA_default", 
                       ref_col = "celltype",
                       palette = c("#fb9a99", "#b2df8a", "#1f78b4", "#d62728", 
                                   "#6fadfd", "#E4D00A", "#FF5733"))
```

# Exploring the data after CIA {#aftercia}

Since CIA seamlessly work on SingleCellExperiment objects, its output is fully compatible with the `Biocpkg("iSEE")` package for interactive data exploration.

Using iSEE on the output of CIA can be as simple as executing this chunk:

```{r launchisee, eval=FALSE}
sce <- Seurat::as.SingleCellExperiment(SO)

if (require(iSEE)) {
  iSEE(sce)
}
```

# Conclusion {#conclusion}

In this notebook, we have showed the robustness and versatility of the **Cluster Independent Annotation (CIA)** method. 
Through the computation of simple signature scores and subsequent cell classification, **CIA provides a fast and efficient approach to cell type annotation in single-cell RNA sequencing data**.
Additionally, we have shown that by increasing the similarity threshold, CIA can effectively mitigate misclassifications, highlighting the presence of ambiguous or novel cell types.
In a landscape where signature files can be generated with increasing ease (e.g. extracted from larger cell atlases), and/or refined over different iteration of the analysis, such a tool can prove to be essential to simplify and optimize this workflow step.

# Session info {- .smaller}

```{r sessioninfo}
sessionInfo()
```

# References
